extensions/elasticsearch/elasticsearch-7/src/main/java/com/hazelcast/jet/elastic/ElasticSourceBuilder.java

```diff
@@ -0,0 +1,238 @@
+        ElasticSourceBuilder<T_NEW> newThis = (ElasticSourceBuilder<T_NEW>) this;
+        newThis.mapToItemFn = checkSerializable(mapToItemFn, "mapToItemFn");
+        return newThis;
+    }
+
+    /**
+     * Set the function that provides {@link RequestOptions}
+     * <p>
+     * It can either return a constant value or a value based on provided request.
+     * <p>
+     * For example, use this to provide a custom authentication header:
+     * <pre>{@code
+     * sourceBuilder.optionsFn((request) -> {
+     *     RequestOptions.Builder builder = RequestOptions.DEFAULT.toBuilder();
+     *     builder.addHeader("Authorization", "Bearer " + TOKEN);
+     *     return builder.build();
+     * })
+     * }</pre>
+     *
+     * @param optionsFn function that provides {@link RequestOptions}
+     * @see <a
+     * href="https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/java-rest-low-usage-requests.html">
+     * RequestOptions in Elastic documentation</a>
+     */
+    @Nonnull
+    public ElasticSourceBuilder<T> optionsFn(@Nonnull FunctionEx<? super ActionRequest, RequestOptions> optionsFn) {
+        this.optionsFn = checkSerializable(optionsFn, "optionsFn");
+        return this;
+    }
+
+    /**
+     * Enable slicing
+     * <p>
+     * Number of slices is equal to globalParallelism (localParallelism * numberOfNodes)
+     * <p>
+     * Use this option to read from multiple shards in parallel.
+     *
+     * @see
+     * <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html#sliced-scroll">
+     *     Sliced Scroll</a>
+     */
+    @Nonnull
+    public ElasticSourceBuilder<T> enableSlicing() {
+        this.slicing = true;
+        return this;
+    }
+
+    /**
+     * Enable co-located reading
+     *
+     * Jet cluster member must run exactly on the same nodes as Elastic cluster.
+     */
+    @Nonnull
+    public ElasticSourceBuilder<T> enableCoLocatedReading() {
+        this.coLocatedReading = true;
+        return this;
+    }
+
+    /**
+     * Set the keepAlive for Elastic search scroll
+     * <p>
+     * See {@link SearchRequest#scroll(String)}
+     *
+     * @param scrollKeepAlive keepAlive value, this must be high enough to process all results from a single scroll,
+     *                        default value 1m
+     */
+    @Nonnull
+    public ElasticSourceBuilder<T> scrollKeepAlive(@Nonnull String scrollKeepAlive) {
+        this.scrollKeepAlive = requireNonNull(scrollKeepAlive, scrollKeepAlive);
+        return this;
+    }
+
+    /**
+     * Set the local parallelism of this source.
+     *
+     * Note this has an effect only when {@link #enableCoLocatedReading()} or {@link #enableSlicing()} is enabled,
+     * otherwise the data is read only from 1 processor in whole Jet cluster.
+     */
+    @Nonnull
+    public ElasticSourceBuilder<T> preferredLocalParallelism(int preferredLocalParallelism) {
```

Comment:
this should not be an option here. The user can set their own and then when slicing is enabled we should set it to reasonable default. Sinks themselves don't have a parallelism option, it's something you add when you add it to the pipeline.